service/history/shard/task_request_tracker.go

```diff
@@ -0,0 +1,171 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package shard
+
+import (
+	"sync"
+
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type (
+	taskRequestCompletionFn func(error)
+
+	taskRequestTracker struct {
+		sync.Mutex
+
+		// using priority queue to track the min pending task key
+		// might be an overkill since the max length of the nested map
+		// is equal to shardIO concurrency limit which should be small
+		pendingTaskKeys      map[tasks.Category]map[tasks.Key]struct{}
+		inflightRequestCount int
+
+		waitChannels []chan<- struct{}
+	}
+)
+
+func newTaskRequestTracker() *taskRequestTracker {
+	outstandingTaskKeys := make(map[tasks.Category]map[tasks.Key]struct{})
+	for _, category := range tasks.GetCategories() {
+		outstandingTaskKeys[category] = make(map[tasks.Key]struct{})
+	}
+	return &taskRequestTracker{
+		pendingTaskKeys: outstandingTaskKeys,
+		waitChannels:    make([]chan<- struct{}, 0),
```

Comment:
any difference to init with 0 length and set it to nil at L170?

Reply:
No I think. Not sure what I am thinking when writing this. Removed.